File: recursion/10/sol.c
#include <stdio.h>
#include <assert.h>
#include <time.h>

// Function declaration
bool dp(int i, int j, const char* s, const char* p, bool** memo, int sLen, int pLen);
bool isMatch(char* s, char* p);


void runTest(char* s, char* p, bool expected, int testCase) {
    printf("Test case %d:\n", testCase);
    printf("Input: s = \"%s\", p = \"%s\"\n", s, p);
    printf("Expected: %s\n", expected ? "true" : "false");
    
    bool result = isMatch(s, p);
    printf("Got: %s\n", result ? "true" : "false");
    
    assert(result == expected);
    printf("Test case %d passed!\n\n", testCase);
}

void runPerformanceTest() {
    printf("Running performance test...\n");
    
    // Create a large input string and pattern
    char s[1001] = {0};
    char p[1001] = {0};
    memset(s, 'a', 1000);
    memset(p, 'a', 500);
    strcat(p, ".*");
    
    clock_t start = clock();
    bool result = isMatch(s, p);
    clock_t end = clock();
    double time_spent = (double)(end - start) / CLOCKS_PER_SEC;
    
    printf("Performance test completed in: %f seconds\n\n", time_spent);
}

int main() {
    // Run performance test
    runPerformanceTest();
    
    // Test case 1: Basic non-matching case
    runTest("aa", "a", false, 1);
    
    // Test case 2: Star pattern
    runTest("aa", "a*", true, 2);
    
    // Test case 3: Dot star pattern
    runTest("ab", ".*", true, 3);
    
    // Test case 4: Complex pattern
    runTest("mississippi", "mis*is*ip*.", true, 4);
    
    // Test case 5: Empty strings
    runTest("", "", true, 5);
    
    printf("All test cases passed!\n");
    return 0;
}

File: recursion/10/sol.py
from my import Solution

if __name__ == "__main__":
    solution = Solution()
    
    # Test case 1: Basic non-matching case
    s1, p1 = "aa", "a"
    result1 = solution.isMatch(s1, p1)
    print(f"Test case 1:")
    print(f"Input: s = {s1}, p = {p1}")
    print(f"Expected: False")
    print(f"Got: {result1}")
    assert result1 == False, "Test case 1 failed"
    
    # Test case 2: Star pattern
    s2, p2 = "aa", "a*"
    result2 = solution.isMatch(s2, p2)
    print(f"\nTest case 2:")
    print(f"Input: s = {s2}, p = {p2}")
    print(f"Expected: True")
    print(f"Got: {result2}")
    assert result2 == True, "Test case 2 failed"
    
    # Test case 3: Dot star pattern
    s3, p3 = "ab", ".*"
    result3 = solution.isMatch(s3, p3)
    print(f"\nTest case 3:")
    print(f"Input: s = {s3}, p = {p3}")
    print(f"Expected: True")
    print(f"Got: {result3}")
    assert result3 == True, "Test case 3 failed"
    
    # Test case 4: Complex pattern
    s4, p4 = "mississippi", "mis*is*ip*."
    result4 = solution.isMatch(s4, p4)
    print(f"\nTest case 4:")
    print(f"Input: s = {s4}, p = {p4}")
    print(f"Expected: True")
    print(f"Got: {result4}")
    assert result4 == True, "Test case 4 failed"
    
    # Test case 5: Empty strings
    s5, p5 = "", ""
    result5 = solution.isMatch(s5, p5)
    print(f"\nTest case 5:")
    print(f"Input: s = {s5}, p = {p5}")
    print(f"Expected: True")
    print(f"Got: {result5}")
    assert result5 == True, "Test case 5 failed"
    
    print("\nAll test cases passed!")